ONSdigital/The-Train
# wired up new file part supplier bringing the previous changes togethe…

### What
[Trello card ](https://trello.com/c/LETwHseQ/4940-fix-the-train-threads-file-stuff-1d)

Replacing & deleted old `FileUploadHelper` with in new `CloseablePartSupplier` (introduced in previous commit). This change wires together the changes introduced the last several commits and should fix the thread leak issue identified in testing.

The thread leak was caused by an issues in the Apache commons `FileUpload` library which was used by this application to handle file upload requests. 

This library has now been removed and replaced with a new class that use object from the Java standard library. The replacement code follows the [example](https://github.com/tipsy/spark-file-upload/blob/master/src/main/java/UploadExample.java) outlined in the [SparkJava documentation](https://sparkjava.com/documentation#examples-and-faq).

#### Before
Testing publishing locally before these changes were implemented. You can see that the number of live threads increases continuously over time.
<img width="605" alt="Screenshot 2020-11-11 at 12 36 32 copy" src="https://user-images.githubusercontent.com/14196957/99994490-fa31e500-2db0-11eb-92f4-45febe34084f.png">

#### After
Testing publishing locally using the code in this PR. The overall live thread count is now much lower, it decreases in the post publish period and stays within a consistent range. I believe this demonstrates the thread leak has now been fixed.
<img width="1410" alt="Screenshot 2020-11-23 at 17 12 28" src="https://user-images.githubusercontent.com/14196957/99994649-2f3e3780-2db1-11eb-88f0-68eac6a3ba6d.png">

### How to review
Code review. 
If you want to test this change locally:
- Run the website publishing stack locally.
- Run `jvisualvm` to start the Java monitoring GUI and open your `The-Train` Java process from the list.
- Create & publishing a collection containing a time series dataset (preferrable a big one - UKEA, LMS).
- Observe the thread behaviour via the monitoring tool. You should see behaviour like the After graph above.
### Who can review

Anyone.
